cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java

```diff
@@ -0,0 +1,221 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.events;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.lang.ClassPathResources;
+import io.cdap.cdap.common.lang.FilterClassLoader;
+import io.cdap.cdap.extension.AbstractExtensionLoader;
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramReference;
+import io.cdap.cdap.spi.events.EventReader;
+import io.cdap.cdap.spi.events.EventResult;
+import io.cdap.cdap.spi.events.StartProgramEvent;
+import io.cdap.cdap.spi.events.StartProgramEventDetails;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.twill.api.RunId;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodically checks event reader for StartProgramEvents. If any events available, publish them to
+ * TMS.
+ */
+public class StartProgramEventHandler extends EventReaderHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StartProgramEventHandler.class);
+  private Collection<EventReader<StartProgramEvent>> readers;
+  private final CConfiguration cConf;
+  private final EventReaderProvider<StartProgramEvent> extensionProvider;
+  private final ProgramLifecycleService lifecycleService;
+  private ScheduledExecutorService executor;
+
+  private ThreadPoolExecutor threadPoolExecutor;
+
+  /**
+   * Create instance that handles StartProgramEvents.
+   *
+   * @param cConf CDAP configuration
+   * @param extensionProvider eventReaderProvider for StartProgramEvent Readers
+   * @param lifecycleService to publish start programs to TMS
+   */
+  @Inject
+  StartProgramEventHandler(CConfiguration cConf, EventReaderProvider<StartProgramEvent> extensionProvider,
+                           ProgramLifecycleService lifecycleService) {
+    this.cConf = cConf;
+    this.extensionProvider = extensionProvider;
+    this.lifecycleService = lifecycleService;
+    threadPoolExecutor = null;
+  }
+
+  @Override
+  public boolean initialize() {
+    readers = extensionProvider.loadEventReaders().values();
+    if (readers.isEmpty()) {
+      return false;
+    }
+    threadPoolExecutor = new ThreadPoolExecutor(readers.size(), readers.size(), 60,
```

Comment:
When is the executor being shutdown? If you don't shutdown it down, it can hang the process for up to 60 seconds.

Reply:
Shutting down in shutdown method now